runtime/lua/vim/diagnostic.lua

```diff
@@ -1165,7 +1142,137 @@ function M.enable(bufnr, namespace)
   end
 end
 
--- }}}
+--- Parse diagnostics from a string.
+---
+--- For example, consider a line of output from a linter:
+--- <pre>
+--- WARNING filename:27:3: Variable 'foo' does not exist
+--- </pre>
+--- This can be parsed into a diagnostic |diagnostic-structure|
+--- with:
+--- <pre>
+--- local s = "WARNING filename:27:3: Variable 'foo' does not exist"
+--- local pattern = "^(%w+) %w+:(%d+):(%d+): (.+)$"
+--- local groups = {"severity", "lnum", "col", "message"}
+--- vim.diagnostic.match(s, pattern, groups, {WARNING = vim.diagnostic.WARN})
+--- </pre>
+---
+---@param str string String to parse diagnostics from.
+---@param pat string Lua pattern with capture groups.
+---@param groups table List of fields in a |diagnostic-structure| to
+---                    associate with captures from {pat}.
+---@param severity_map table A table mapping the severity field from {groups}
+---                          with an item from |vim.diagnostic.severity|.
+---@param defaults table|nil Table of default values for any fields not listed in {groups}.
+---                          When omitted, numeric values default to 0 and "severity" defaults to
+---                          ERROR.
+---@return diagnostic |diagnostic-structure|
+function M.match(str, pat, groups, severity_map, defaults)
+  vim.validate {
+    str = { str, 's' },
+    pat = { pat, 's' },
+    groups = { groups, 't' },
+    severity_map = { severity_map, 't', true },
+    defaults = { defaults, 't', true },
+  }
+
+  defaults = defaults or {}
+  severity_map = severity_map or M.severity
+
+  local diagnostic = {}
+  for i, match in ipairs({string.match(str, pat)}) do
+    local field = groups[i]
+    if field == "severity" then
+      match = severity_map[match]
+    elseif field == "lnum" or field == "end_lnum" or field == "col" or field == "end_col" then
+      match = assert(tonumber(match)) - 1
+    end
+    diagnostic[field] = match
+  end
+  diagnostic.severity = diagnostic.severity or defaults.severity or vim.severity.ERROR
+  diagnostic.col = diagnostic.col or 0
+  diagnostic.end_lnum = diagnostic.end_lnum or diagnostic.lnum
+  diagnostic.end_col = diagnostic.end_col or diagnostic.col
+  return diagnostic
+end
+
+--- Parse diagnostics from an error format.
+---
+---@param efm string Error format |errorformat|.
+---@param lines table List of lines to parse
+---@return array of diagnostics |diagnostic-structure|
+function M.from_errorformat(efm, lines)
+  local qflist = vim.fn.getqflist({efm = efm, lines = lines})
+  return M.fromlist(qflist.items)
+end
```

Comment:
Not sure if it is worth to expose this given that it only takes two lines to implement, but no strong opinion on it.

Reply:
Fair point, this would also save me from writing tests for it 😄